orttraining/orttraining/lazy_tensor/accelerator.cpp

```diff
@@ -0,0 +1,488 @@
+  ExampleRun(args);
+  DynamicSettings::GetInstance().SetOnnxFusionFlag(true);
+  // Storage of compilation.
+  CompiledObject compiled;
+  // Create an empty session.
+  compiled.sess = CreateSession();
+  // Let's get the empty session and initialize it.
+  onnxruntime::InferenceSession& sess = *compiled.sess;
+  // Export subgraph_ to ONNX.
+  const std::string serialized_model = ExportToOnnx(subgraph_, args);
+  // Memory info for all tensors.
+  // Assume all inputs are on the same device.
+  OrtDevice shared_device = CheckAndGetTensorDevice(args);
+  // Load ONNX model into session, register
+  // EPs and finally initialize session.
+  InitializeSession(shared_device, serialized_model, sess);
+  // Clean model file.
+  // ORT_ENFORCE(std::remove(model_path.c_str()) == 0, "Failed to remove temporary file: ", model_path);
+
+  onnxruntime::RunOptions run_options;
+  std::vector<std::string> feed_names;
+  std::vector<std::string> fetch_names;
+
+  for (auto node_arg : *sess.GetModelInputs().second) {
+    feed_names.push_back(node_arg->Name());
+  }
+  for (auto node_arg : *sess.GetModelOutputs().second) {
+    fetch_names.push_back(node_arg->Name());
+  }
+
+  // Duplicate device info for putting output tensors on the shared device.
+  std::vector<OrtDevice> fetches_device_info(fetch_names.size(), shared_device);
+
+  // Create a callable which feeds inputs to ORT
+  // session's Run(...) and returns outputs.
+  auto code = [this, run_options,
+               feed_names, fetch_names,
+               fetches_device_info, &sess](at::ArrayRef<c10::IValue>& args) {
+    // Inputs of ORT session.
+    std::vector<OrtValue> feeds;
+    // Outputs of ORT session.
+    std::vector<OrtValue> fetches;
+
+    {
+#ifdef USE_CUDA
+      NvtxRange range("Prepare inputs");
+#endif
+      // Prepare inputs.
+      const auto num_inputs = subgraph_->inputs().size();
+      for (size_t i = 0; i < num_inputs; ++i) {
+        // The value can be either tensor or scalar.
+        // Scalar is a tensor with empty shape vector.
+        // Create ORT tensor from Pytorch tensor without copy.
+        if (args.at(i).isScalar()) {
+          // Scalar.
+          // ORT_ENFORCE(subgraph_->inputs().at(i)->type()->kind() == c10::TypeKind::TensorType);
+          feeds.push_back(CreateOrtScalarValue(args.at(i).toScalar()));
+        } else if (args.at(i).isTensor()) {
+          // Tensor.
+          ORT_ENFORCE(subgraph_->inputs().at(i)->type()->kind() == c10::TypeKind::TensorType);
+          feeds.push_back(CreateOrtTensorValue(args.at(i).toTensor()));
+        } else {
+          // Looks like LTC only passes scalars and tensors into backend, so we don't care
+          // other types for now.
+          ORT_THROW("Only tensor inputs are supported.");
+        }
+      }
+    }
+
+    {
+#ifdef USE_CUDA
+      NvtxRange range("Call sess.Run");
+#endif
+      // Inputs are ready. Let's run ORT.
+      ORT_THROW_IF_ERROR(sess.Run(
```

Comment:
one thing i guess we haven't pay much attention to yet: how did we make sure we are on the same stream with the pytorch? and if user defined customized streams in pytorch, like "with torch.cuda.stream", how will this reflect to our backend?

Reply:
The `current` stream from Pytorch can be retrieved by calling [this function](https://pytorch.org/cppdocs/api/structc10_1_1cuda_1_1_c_u_d_a_stream_guard.html?highlight=current_stream#_CPPv4NK3c104cuda15CUDAStreamGuard14current_streamEv). ORT session or Aaron's new API should consume this stream.